### PR TITLE
[autoscaler][kuberay] Deflake KubeRay autoscaling test

### LINKED ISSUE #26377: [KubeRay][CI] KubeRay autoscaling test is flakey
### What happened + What you expected to happen

The KubeRay autoscaling test is flaky. 
It's on this list.
https://flakey-tests.ray.io/#showAll=true

That should be fixed -- we want high confidence in this feature.

### Versions / Dependencies

Ray master

### Reproduction script

N/A

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
## Why are these changes needed?

Deflakes the KubeRay autoscaling e2e test, which appears to be suffering from a race condition involving CRD creation and registration. 
Background: registering a CRD takes a bit of time after the request to create the CRD object returns.

Strategy:

- Move CRD/operator setup/teardown out of the test and into the test pipeline. This avoid race conditions related to infra creation and deletion during test retries.
- After creating the CRD and verifying the CRD exists, try a few times to create a test CR.
  * Once that's done, we're confident the CRD is registered and the test can proceed. 

## Related issue number

Closes https://github.com/ray-project/ray/issues/26377

## Checks

I ran the test 12 times in the CI with these changes and didn't observe any failures.

### CODE REVIEW

File: python/ray/tests/kuberay/test_autoscaling_e2e.py [hunk 1]
```diff
@@ -33,26 +33,18 @@
...
 logger = logging.getLogger(__name__)
-logging.basicConfig(
-    level=logging.INFO,
-    format="[%(levelname)s %(asctime)s] " "%(filename)s: %(lineno)d  " "%(message)s",
-)
 
 # This image will be used for both the Ray nodes and the autoscaler.
 # The CI should pass an image built from the test branch.
-RAY_IMAGE = os.environ.get("RAY_IMAGE", "rayproject/ray:448f52")
+RAY_IMAGE = os.environ.get("RAY_IMAGE", "rayproject/ray:nightly-py38")
```
Comment (DmitriGekhtman) [inline python/ray/tests/kuberay/test_autoscaling_e2e.py:39]: No particular reason for this default besides the fact that I use a py38 environment locally.

Comment (DmitriGekhtman) [reply python/ray/tests/kuberay/test_autoscaling_e2e.py:39]: (The CI specifies and image built from the PR branch.)

**Comment by brucez-anyscale** [comment]
This pr looks good. But why move the code outside the python to pipeline will help?
Because pipeline is slower and have wait time between steps?

**Comment by DmitriGekhtman** [comment]
> This pr looks good. But why move the code outside the python to pipeline will help? Because pipeline is slower and have wait time between steps?

I should clarify that -- the pipeline is set up to retry tests. I think that this external retry logic may introduce race conditions around creation and teardown of the operator and CRD.
To avoid this, I'd prefer to set up the CRD and operator as part of the test's Kubernetes infrastructure (similar to the KinD setup).

File: .buildkite/pipeline.yml [hunk 2]
```diff
@@ -442,13 +442,17 @@
...
+    - python python/ray/tests/kuberay/setup/setup_kuberay.py
     - echo "--- Running the test."
     - bazel test --config=ci $(./ci/run/bazel_export_options)
       --test_tag_filters=kuberay_operator
       --test_env=RAY_IMAGE=docker.io/library/ray-ci:kuberay-test
       --test_env=PULL_POLICY=IfNotPresent
       --test_env=KUBECONFIG=/root/.kube/config
       python/ray/tests/...
+    - echo "--- Tearing down KubeRay operator."
+    - python python/ray/tests/kuberay/setup/teardown_kuberay.py
```
Comment (simon-mo) [inline .buildkite/pipeline.yml:455]: this should probably go into the cleanup clause (first row in list) and made idempotent so it's cleaned up regardless

**Review by simon-mo — APPROVED** [review]

File: .buildkite/pipeline.yml (hunk 2 above)
Comment (DmitriGekhtman) [reply .buildkite/pipeline.yml]: Added --ignore-not-found to kubectl delete to make it idempotent.
Tacked it on to the end of the cleanup function.

**Review by brucez-anyscale — APPROVED** [review]

**Comment by brucez-anyscale** [comment]
looks like multiple test failure should be fixed.

**Comment by DmitriGekhtman** [comment]
Not sure how these changes could have triggered the failures, but that's quite a few failures.

**Comment by DmitriGekhtman** [comment]
> Not sure how these changes could have triggered the failures, but that's quite a few failures.

It's looking better after rebasing.

**Pull request merged**